fdbserver/ClusterController.actor.cpp

```diff
@@ -516,33 +463,160 @@ class ClusterControllerData {
-			    .detail("NumExclusionWorkers", exclusionWorkerIds.size())
-			    .detail("PolicyGenerations", SERVER_KNOBS->POLICY_GENERATIONS)
-			    .backtrace();
+			    .detail("NumExclusionWorkers", exclusionWorkerIds.size());
 
 			logServerSet->clear();
 			logServerSet.clear();
 			throw no_more_servers();
 		}
 
+		// If we have less than the desired amount, return all of the processes we have
+		if (logServerSet->size() <= desired) {
+			for (auto& object : logServerMap->getObjects()) {
+				results.push_back(*object);
+			}
+			for (auto& result : results) {
+				id_used[result.interf.locality.processId()]++;
+			}
+			TraceEvent("GetTLogTeamDone")
+			    .detail("DcIds", dcList)
+			    .detail("Policy", policy->info())
+			    .detail("Results", results.size())
+			    .detail("Processes", logServerSet->size())
+			    .detail("Workers", id_worker.size())
+			    .detail("Required", required)
+			    .detail("Desired", desired)
+			    .detail("Fitness", requiredFitness)
+			    .detail("Used", requiredUsed)
+			    .detail("AddingDegraded", requiredDegraded)
+			    .detail("InCCDC", requiredInCCDC);
+			return results;
+		}
+
+		// If we have added any degraded processes, try and remove them to see if we can still
+		// have the desired amount of processes
+		if (requiredDegraded) {
+			logServerMap->clear();
+			for (auto workerIter = fitness_workers.begin(); workerIter != fitness_workers.end(); ++workerIter) {
+				auto fitness = std::get<0>(workerIter->first);
+				auto used = std::get<1>(workerIter->first);
+				if (fitness > requiredFitness || (fitness == requiredFitness && used > requiredUsed)) {
+					break;
+				}
+				auto addingDegraded = std::get<2>(workerIter->first);
+				if (addingDegraded) {
+					continue;
+				}
+				for (auto& worker : workerIter->second) {
+					logServerMap->add(worker.interf.locality, &worker);
+				}
+			}
+			if (logServerSet->size() >= desired && logServerSet->validate(policy)) {
+				requiredDegraded = false;
+			}
+		}
+
+		// If we have added any processes in the CC DC, try and remove them to see if we can still
+		// have the desired amount of processes
+		if (requiredInCCDC) {
+			logServerMap->clear();
+			for (auto workerIter = fitness_workers.begin(); workerIter != fitness_workers.end(); ++workerIter) {
+				auto fitness = std::get<0>(workerIter->first);
+				auto used = std::get<1>(workerIter->first);
+				if (fitness > requiredFitness || (fitness == requiredFitness && used > requiredUsed)) {
+					break;
+				}
+				auto addingDegraded = std::get<2>(workerIter->first);
+				auto inCCDC = std::get<3>(workerIter->first);
+				if (inCCDC || (!requiredDegraded && addingDegraded)) {
+					continue;
+				}
+				for (auto& worker : workerIter->second) {
+					logServerMap->add(worker.interf.locality, &worker);
+				}
+			}
+			if (logServerSet->size() >= desired && logServerSet->validate(policy)) {
+				requiredInCCDC = false;
+			}
+		}
+
+		logServerMap->clear();
```

Comment:
Nit: maybe also add a comment here as why we build the `logServerMap` the forth time in this function (my understanding is that now we know that whether we can form a tlog set without adding degraded workers and workers in the same DC as CC, so now we for sure can find enough `fit` workers for the tlog set, correct?).